### PR TITLE
[pull-request:create] [WIP] Check for commits between local branch and remote base before push

### PR DESCRIPTION
**Pull request by phansys**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | kinda
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

TODO:
- [x] Add tests.

### CODE REVIEW

**Comment by cordoval** [comment]
@phansys it is now i think a matter of updating the expectations on the tests and getting this fixed, thanks great job

**Comment by phansys** [comment]
I'd like to know which could be the right behavior, because currently the PR is opened based on commits from remote origin branch, regardless it is in sync with the local. This leads to situations where two branches (local and remote origin) can share the same name but they also can have two completely separated histories. So, I need some concerns of you about how to proceed in these cases. I see 3 possible scenarios, but IMO the first could be better:

1. **Combining local branch and remote origin (if exists):**
  Check for commits between local branch and target base, if there are commits, try to push them to the user's remote (origin), regardless these commits are already present on the remote. If the push fails, it means that something went wrong on that history, so the user must fix the situation by hand. If the push went well, proceed with the PR creation.
2. **Force local branch to override the remote origin (if exists):**
 Check for commits between local branch and target base, if there are commits, do a forced push (`git push user branch --force`) to the user's remote (origin), regardless these commits are already present on the remote. If the push fails, it means that something went wrong, so the user must fix the situation by hand. If the push went well, proceed with the PR creation.
3. **Check for commits between local branch and remote origin (if exists):**
 Check for commits between local branch and target base, if there are commits, check if they are the same that are present at the user's remote (origin), if not, fail indicating that situation must be fixed by hand.

Please, feel free to propose any other solution.

**Comment by cordoval** [comment]
these are atomic actions

it is always the case that it is assumed they are in sync, it is responsibility to call before gush b:p -u cordoval -f or just using git to force if need be. Let's keep things simple.

**Comment by phansys** [comment]
Can I do a little update in this PR in order to use a new message for the PR creation?
```
$ gush pull-request:create

Open request on gushphp/gush
============================

 phansys (You) want to merge 1 commit into gushphp/gush:master from phansys:check_for_commits_before_push.
```
image

This is just a suggestion in order to leverage the check done against the local commits, while using a similar wording as GitHub. I don't know if this can be confusing here, since we are proposing the PR at this step, not merging it.
Any thoughts?

**Comment by cordoval** [comment]
that is correct

File: src/Command/PullRequest/PullRequestCreateCommand.php [hunk 1]
```diff
@@ -148,13 +148,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-        $this->guardCommitsBetweenLocalBranchAndRemoteBaseBranch($org, $repo, $base, $sourceBranch);
+        $commitCount = $this->guardCommitsBetweenLocalBranchAndRemoteBaseBranch($org, $repo, $base, $sourceBranch);
 
         $styleHelper->title(sprintf('Open request on %s/%s', $org, $repo));
+
         $styleHelper->text(
             [
-                sprintf('This pull-request will be opened on "%s/%s:%s".', $org, $repo, $base),
-                sprintf('The source branch is "%s:%s".', $sourceOrg, $sourceBranch),
+                sprintf('%1$s (You) want to merge %2$u %7$s into <comment>%3$s/%4$s:%5$s</comment> from <comment>%1$s:%6$s</comment>.', $sourceOrg, $commitCount, $org, $repo, $base, $sourceBranch, $commitCount > 1 ? 'commits' : 'commit'),
```
Comment (cordoval) [inline src/Command/PullRequest/PullRequestCreateCommand.php]: leave this as just the username and do wants instead of want

**Comment by phansys** [comment]
The main goal of this PR is to not push any branch to user's remote if there aren't commits between the local and the base branch. What do you mean?
image

**Comment by cordoval** [comment]
LGTM i will review  further when i come back from the park, going with my daughter 💃

**Comment by sstok** [comment]
You can check when the remote branch exists, the remote HEAD equals the HEAD of the local branch, and simple warn if they don't: "The remote branch exists but doesn't equal your remote branch! This may cause problems. Please check this is correct before you continue {commits of local} {commits and remote}".

Doing actions automatically should only be done when it's safe (no magic), that's why [PullRequestCheckoutCommand](https://github.com/gushphp/gush/blob/master/src/Command/PullRequest/PullRequestCheckoutCommand.php#L71) checks if the branch already exists and errors when the remote doesn't equal.

`pull-request:create` does a push because it's something you easily forget, and then find out afterwards (having to type everything again 💀 ).

**Comment by phansys** [comment]
I think I can play with any of the proposed paths, continue to delegate these responsibilities to other commands (the current behavior) or add some extra checks for these scenarios like proposed before. I'm awaiting for your deliberation about how to solve these possible situations.
Thank you guys.

**Pull request merged**